seqr/views/apis/igv_api.py

```diff
@@ -126,9 +132,38 @@ def fetch_igv_track(request, project_guid, igv_track_path):
     if igv_track_path.endswith('.bam.bai') and not does_file_exist(igv_track_path, user=request.user):
         igv_track_path = igv_track_path.replace('.bam.bai', '.bai')
 
+    if is_google_bucket_file_path(igv_track_path):
+        return _stream_gs(request, igv_track_path)
+
     return _stream_file(request, igv_track_path)
 
 
+def _stream_gs(request, gs_path):
+    headers = {'Authorization': 'Bearer {}'.format(_get_access_token(request.user))}
+    range_header = request.META.get('HTTP_RANGE')
+    if range_header:
+        headers['Range'] = range_header
+
+    bucket_object = gs_path.replace('gs://', '', 1).split('/', 1)
+    response = requests.get(
+        'https://{bucket}.storage.googleapis.com/{object}'.format(bucket=bucket_object[0], object=bucket_object[1]),
+        headers=headers,
+        stream=True)
+
+    return StreamingHttpResponse(response.iter_content(chunk_size=65536), status=response.status_code,
+                                 content_type='application/octet-stream')
+
+
+def _get_access_token(user):
+    access_token = safe_redis_get_json(IGV_CACHE_KEY)
+    if not access_token or (time.time() - access_token['timer']) > EXPIRATION_TIME_IN_SECONDS:
+        process = run_command('gcloud auth print-access-token', user=user)
+        if process.wait() == 0:
+            access_token = {'timer': time.time(), 'token': next(process.stdout).decode('utf-8').strip()}
+            safe_redis_set_json(IGV_CACHE_KEY, access_token)
```

Comment:
Like I explained in previous reviews, redis allows you to set key/values that have an expiration. You should not be storing expiration numbers in json in setting them and checking them. You should be setting your cache with an explicit expiration so REDIS will handle deleting things that are expire - 
```
access_token = safe_redis_get_json(GS_STORAGE_ACCESS_TOKEN)
if not access_token:
    process = run_command('gcloud auth print-access-token', user=user)
    if process.wait() == 0:
        access_token = next(process.stdout).decode('utf-8').strip()
        safe_redis_set_json(GS_STORAGE_ACCESS_TOKEN, access_token, expire=EXPIRATION_TIME_IN_SECONDS)
return access_token
```

Reply:
> Awesome work! One thing I want to double check, this will change how all IGV files are handled. Did you check to see that crams, gcnv, and rnaseq all still work as expected?

@ShifaSZ any answer for this?

Reply:
I have just double-checked that the alignment, gcnv, and rnaseq work fine. They fetch track data from the GS.